@utahdts/utah-design-system/react/components/forms/FileInput/FileInput.jsx

```diff
@@ -0,0 +1,171 @@
+import { useImmer } from 'use-immer';
+import {
+  useCallback, useEffect, useRef
+} from 'react';
+import { isEqual, isFunction } from 'lodash';
+import { RequiredStar } from '../RequiredStar';
+import { joinClassNames } from '../../../util/joinClassNames';
+import { Tag } from '../../buttons/Tag';
+import { useAriaMessaging } from '../../../contexts/UtahDesignSystemContext/hooks/useAriaMessaging';
+import { ErrorMessage } from '../ErrorMessage';
+import { FileContext } from './context/FileContext';
+
+/**
+ * @param {object} props
+ * @param {import('react').ReactNode} [props.children]
+ * @param {string} [props.acceptedFileTypes]
+ * @param {string} [props.className]
+ * @param {string} [props.errorMessage]
+ * @param {string} [props.hint]
+ * @param {string} props.id
+ * @param {import('react').Ref<HTMLDivElement>} [props.innerRef]
+ * @param {string} props.label
+ * @param {boolean} [props.multiple]
+ * @param {string} [props.name]
+ * @param {boolean} [props.isDisabled]
+ * @param {boolean} [props.isRequired]
+ * @param {import('react').ChangeEventHandler} [props.onChange]
+ * @param {FileList} [props.value]
+ * @returns {import('react').JSX.Element}
+ */
+export function FileInput({
+  acceptedFileTypes,
+  children,
+  className,
+  errorMessage,
+  hint,
+  id,
+  innerRef,
+  isDisabled,
+  isRequired,
+  label,
+  multiple,
+  name,
+  onChange,
+  value,
+}) {
+  const { addPoliteMessage } = useAriaMessaging();
+  const [isDragged, setDragged] = useImmer(false);
+  const [files, setFiles] = useImmer(value || []);
+  const inputRef = useRef(null);
+
+  const currentOnChange = useCallback(() => {
+    if (onChange) {
+      // @ts-ignore
+      onChange(inputRef.current.files);
+    } else {
+      // @ts-ignore
+      setFiles(inputRef.current.files);
+    }
+  }, []);
+
+  const removeFile = useCallback((/** @type {File} */ file) => {
+    // @ts-ignore
+    const currentFiles = [...inputRef.current.files];
+    const fileIndex = currentFiles.findIndex((item) => isEqual(file, item));
+    if (fileIndex !== -1) {
+      currentFiles.splice(fileIndex, 1);
+      // Create new FileList
+      const dataTransfer = new DataTransfer();
+      currentFiles.forEach((item) => dataTransfer.items.add(item));
+      // @ts-ignore
+      inputRef.current.files = dataTransfer.files;
+      currentOnChange();
+    }
+  }, []);
+
+  useEffect(() => {
+    if (files.length) {
+      addPoliteMessage(`You have selected the file: ${[...files].map((item) => item.name).join(', ')}.`);
+    } else {
+      addPoliteMessage('No file selected.');
+    }
+  }, [files]);
+
+  useEffect(() => {
+    // Track files when controlled
+    if (onChange) { setFiles(value || []); }
```

Comment:
this makes sure that if there is an onchange that setFiles also gets called? maybe currentOnChange could be:
```
onChange?.(inputRef.current.files);
setFiles(inputRef.current.files);
```
This keeps onChange and setFiles in sync. Since currentOnChange is called below (and not onChange) then this useEffect could be removed.

Reply:
Need to be tested.